lens/fetcher.go

```diff
@@ -0,0 +1,400 @@
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/fetcher"
+	"github.com/sourcenetwork/defradb/planner/mapper"
+)
+
+// todo: The code in here can be significantly simplified with:
+// https://github.com/sourcenetwork/defradb/issues/1589
+
+type lensedFetcher struct {
+	source   fetcher.Fetcher
+	registry client.LensRegistry
+	lens     Lens
+
+	txn datastore.Txn
+
+	col *client.CollectionDescription
+	// Cache the fieldDescriptions mapped by name to allow for cheaper access within the fetcher loop
+	fieldDescriptionsByName map[string]client.FieldDescription
+
+	targetVersionID string
+}
+
+var _ fetcher.Fetcher = (*lensedFetcher)(nil)
+
+// NewFetcher returns a new fetcher that will migrate any documents from the given
+// source Fetcher as they are are yielded.
+func NewFetcher(source fetcher.Fetcher, registry client.LensRegistry) fetcher.Fetcher {
+	return &lensedFetcher{
+		source:   source,
+		registry: registry,
+	}
+}
+
+func (f *lensedFetcher) Init(
+	col *client.CollectionDescription,
+	fields []client.FieldDescription,
+	filter *mapper.Filter,
+	docmapper *core.DocumentMapping,
+	reverse bool,
+	showDeleted bool,
+) error {
+	f.col = col
+
+	f.fieldDescriptionsByName = make(map[string]client.FieldDescription, len(col.Schema.Fields))
+	// Add cache the field descriptions in reverse, allowing smaller-index fields to overwrite any later
+	// ones.  This should never really happen here, but it ensures the result is consistent with col.GetField
+	// which returns the first one it finds with a matching name.
+	for i := len(col.Schema.Fields) - 1; i >= 0; i-- {
+		field := col.Schema.Fields[i]
+		f.fieldDescriptionsByName[field.Name] = field
+	}
+
+	f.targetVersionID = col.Schema.VersionID
+	return f.source.Init(col, fields, filter, docmapper, reverse, showDeleted)
+}
+
+func (f *lensedFetcher) Start(ctx context.Context, txn datastore.Txn, spans core.Spans) error {
+	history, err := getTargetedSchemaHistory(ctx, txn, f.registry.Config(), f.col.Schema.SchemaID, f.col.Schema.VersionID)
+	if err != nil {
+		return err
+	}
+	f.lens = new(f.registry, f.col.Schema.VersionID, history)
+	f.txn = txn
+
+	return f.source.Start(ctx, txn, spans)
+}
+
+func (f *lensedFetcher) FetchNext(ctx context.Context) (fetcher.EncodedDocument, error) {
+	panic("This function is never called and is dead code.  As this type is internal, panicing is okay for now")
+}
+
+func (f *lensedFetcher) FetchNextDecoded(ctx context.Context) (*client.Document, error) {
+	doc, err := f.source.FetchNextDecoded(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if doc == nil {
+		return nil, nil
+	}
+
+	if doc.SchemaVersionID == f.targetVersionID {
+		// If the document is already at the target schema version, no migration is required and
+		// we can return it early.
+		return doc, nil
+	}
+
+	sourceLensDoc, err := clientDocToLensDoc(doc)
```

Comment:
I tried to use that first, but `ToMap` is not suitable as it only yields rendered items, and had a few other problems.

Reply:
Anything worth mentioning? Would want to know if some parts of the doc API are insufficient, not sure what you mean by rendered items here.

Reply:
> Anything worth mentioning

I think there were little things like deleted/status flags and foreign key field issues. Was more than 2 weeks ago though, so it might have been for different little reasons.

> not sure what you mean by rendered items here.

ToMap only yields fields that the user has requested, not other fields fetched internally for other reasons (e.g. stuff required by aggregates or filters are omitted from the result).

Reply:
but the doc returned by the fetcher doesnt contain any of that, and I dont see why the lens transform would need it. Both the docFetcher, and the lens transformation should just deal with base level fields on the doc.

Reply:
I think you are mistaken here John.  Feel free to call about this if you want me to walk you through.